lib/src/syn/v2/lexer/reader.rs

```diff
@@ -0,0 +1,228 @@
+use thiserror::Error;
+
+use crate::syn::v2::token::Span;
+use std::{fmt, marker::PhantomData, ptr::NonNull};
+
+#[derive(Error, Debug)]
+pub enum CharError {
+	#[error("found eof inside multi byte character")]
+	Eof,
+	#[error("string is not valid utf-8")]
+	Unicode,
+}
+
+#[derive(Clone)]
+pub struct BytesReader<'a> {
+	start: NonNull<u8>,
+	end: NonNull<u8>,
+	current: NonNull<u8>,
+	marker: PhantomData<&'a [u8]>,
+}
+
+impl fmt::Debug for BytesReader<'_> {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.debug_struct("BytesReader")
+			.field("used", &self.used())
+			.field("remaining", &self.remaining())
+			.finish()
+	}
+}
+
+impl<'a> BytesReader<'a> {
+	pub fn new(slice: &'a [u8]) -> Self {
+		let len = slice.len();
+		let start = slice.as_ptr() as *mut u8;
+		// SAFETY: len is derived from the slice so the end pointer is still inside the range for
+		// providance.
+		let end = unsafe { start.add(len) };
+
+		// Pointers derived from references to non-empty sized slice are never null but
+		// I couldn't find information about slices which are empty. Testing seems to indicate
+		// that they will always have a non-null value but I can't be certain of that.
+		let start = NonNull::new(start).unwrap();
```

Comment:
I agree that this seems to be the case however I couldn't find any documentation which guarantees that this is the case so thats why I document it. I didn't assert on empty slices since an empty slice is a valid query.